### PR TITLE
[Fix] attn_dropout

### LINKED ISSUE #122: `BlockSparseAttention` doesn't use dropout
# 🐛 Bug

While working on refactoring the `blocksparse.py` file into its own custom tensor type, I noticed that `BlockSparseAttention` doesn't implement dropout, which is probably not intended.

cc @blefaudeux

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
Fixes #122 + adds a unit test to catch faulty dropouts
@fmassa is rewriting some of that part, but in the meantime `main` should not be knowingly broken

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by blefaudeux** [comment]
@fmassa @dianaml0 turns out that in many case dropout was not applied, not just for blocksparse, a unit test is catching that now :(

**Comment by dianaml0** [comment]
> @fmassa @dianaml0 turns out that in many case dropout was not applied, not just for blocksparse, a unit test is catching that now :(

Oh wow :( Great that you were able to catch that!

**Comment by blefaudeux** [comment]
> > @fmassa @dianaml0 turns out that in many case dropout was not applied, not just for blocksparse, a unit test is catching that now :(
> 
> Oh wow :( Great that you were able to catch that!

maybe a semi-recent change, I think that @fmassa is right in that we should refactor core/. Else for some variant the setting was silently ignored (but not explicitly present in the constructor), so it would just fail for the factory, not as bad but still... Unit tests for the win, this should not happen again

File: xformers/components/attention/core.py [hunk 1]
```diff
@@ -172,10 +173,14 @@ def _apply_dropout(att, dropout):
...
             values = dropout(values)
             att = torch.sparse_coo_tensor(att.indices(), values, att.shape)
+        else:
+            # Simple dense case
+            att = dropout(att)
 
         return att
 
     # Non optimized vanilla dropout
+    print(att - dropout(att))
```
Comment (dianaml0) [inline xformers/components/attention/core.py:183]: Should this be removed?

File: tests/test_attentions.py [hunk 2]
```diff
@@ -110,7 +113,12 @@ def test_order_invariance(
         torch.allclose(results[:, shuffle, :], results_shuffled)
 
         # Test the non-self-attention codepath
-        _ = multi_head(inputs, inputs_shuffled, inputs)
+        att = multi_head(inputs, inputs_shuffled, inputs)
+
+        # Check that dropout actually drops some values
```
Comment (dianaml0) [inline tests/test_attentions.py:118]: Great to have this test now!

File: xformers/components/attention/linformer.py [hunk 3]
```diff
@@ -37,6 +37,7 @@ def __init__(
         if k is None:
             k = seq_len // 4
 
+        print("dropout ", dropout)
```
Comment (dianaml0) [inline xformers/components/attention/linformer.py:40]: Same here?

**Review by dianaml0 — APPROVED** [review]
Really nice catch!

**Pull request merged**